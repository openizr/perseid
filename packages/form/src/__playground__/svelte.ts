/* c8 ignore start */
/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */

import { SvelteComponent } from 'svelte';
import Engine from 'scripts/core/Engine';
import Form from 'scripts/svelte/Form.svelte';
import Field from '__playground__/Field.svelte';

let app: Form;
const { log } = console;

function main(): void {
  app = new Form({
    props: {
      Field: Field as typeof SvelteComponent | null | undefined,
      configuration: {
        root: 'root',
        fields: {
          test: { type: 'string', required: true },
        },
        steps: {
          root: {
            fields: ['test'],
          },
        },
        plugins: [
          (engine: Engine): void => {
            engine.on('userAction', (data, next) => {
              log(data);
              return next(data);
            });
          },
        ],
      },
    },
    target: document.getElementById('root') as unknown as HTMLElement,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Vue components tree.
window.addEventListener('beforeunload', () => {
  (app as { $destroy: () => void; }).$destroy();
});
