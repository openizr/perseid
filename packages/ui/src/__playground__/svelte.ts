/* c8 ignore start */

import type { SvelteComponent } from 'svelte';
import Router from '__playground__/pages/Router.svelte';

let app: SvelteComponent;

function main(): void {
  const target = document.querySelector('#root') as unknown as HTMLElement;
  target.innerHTML = '';
  app = new Router({
    hydrate: false,
    target,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Svelte components tree.
window.addEventListener('beforeunload', () => {
  app.$destroy();
});
