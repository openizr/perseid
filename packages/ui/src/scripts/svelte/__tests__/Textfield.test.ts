/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @vitest-environment jsdom
 */

import UITextfield from 'scripts/svelte/Textfield.svelte';
import { render, fireEvent } from '@testing-library/svelte';

describe('svelte/UITextfield', () => {
  vi.mock('scripts/core/generateRandomId');
  vi.useFakeTimers();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly - basic', () => {
    const { container } = render(UITextfield, { props: { name: 'test', modifiers: 'large' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with id', () => {
    const { container } = render(UITextfield, { props: { name: 'test', id: 'my-id' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with autofocus', () => {
    const { container } = render(UITextfield, { props: { name: 'test', autofocus: true } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with type number', () => {
    const { container } = render(UITextfield, {
      props: {
        name: 'test', type: 'number', min: 0, max: 30, step: 10,
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with left icon', () => {
    const { container } = render(UITextfield, { props: { name: 'test', icon: 'star' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with right icon', () => {
    const { container } = render(UITextfield, {
      props: {
        name: 'test', icon: 'star', iconPosition: 'right',
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with placeholder', () => {
    const { container } = render(UITextfield, { props: { name: 'test', placeholder: 'test...' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with label', () => {
    const { container } = render(UITextfield, { props: { name: 'test', label: '*Label*' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with helper', () => {
    const { container } = render(UITextfield, { props: { name: 'test', helper: '*Helper*' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with value', () => {
    const { container } = render(UITextfield, { props: { name: 'test', value: 'my value' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with disabled', async () => {
    const onChange = vi.fn();
    const { container } = render(UITextfield, { props: { name: 'test', disabled: true, onChange } });
    const input = container.getElementsByTagName('input')[0];
    await fireEvent.input(input, { value: 'new value' });
    expect(container.firstChild).toMatchSnapshot();
    expect(onChange).not.toHaveBeenCalled();
  });

  test('renders correctly - autocomplete off', () => {
    const { container } = render(UITextfield, { props: { name: 'test', autocomplete: 'off' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - readonly', async () => {
    const onChange = vi.fn();
    const { container } = render(UITextfield, { props: { name: 'test', readonly: true, onChange } });
    const input = container.getElementsByTagName('input')[0];
    await fireEvent.keyDown(input, { key: 'A' });
    await fireEvent.keyDown(input, { key: 'a', ctrlKey: true });
    await fireEvent.keyDown(input, { key: 'a', shiftKey: true });
    await fireEvent.keyDown(input, { key: 'a', altKey: true });
    await fireEvent.keyDown(input, { key: 'a', metaKey: true });
    await fireEvent.input(input, { value: 'new value' });
    expect(container.firstChild).toMatchSnapshot();
    expect(onChange).not.toHaveBeenCalled();
  });

  test('renders correctly - with transform and allowedKeys', async () => {
    const onChange = vi.fn();
    const onKeyDown = vi.fn();
    const onPaste = vi.fn();
    const transform = (value: string): [string, number?] => [value.toUpperCase()];
    const { container, rerender } = render(UITextfield, {
      props: {
        name: 'test',
        size: 10,
        transform,
        onChange,
        onKeyDown,
        onPaste,
        allowedKeys: {
          default: /[a-z]/i,
          altKey: /[a-z]/i,
          shiftKey: /[a-z]/i,
          metaKey: /[a-z]/i,
          ctrlKey: /[a-z]/i,
        },
      },
    });
    let input = container.getElementsByTagName('input')[0];
    await fireEvent.keyDown(input, { key: 'A' });
    await fireEvent.keyDown(input, { key: '0' });
    await fireEvent.keyDown(input, { key: 'a', ctrlKey: true });
    await fireEvent.keyDown(input, { key: 'a', shiftKey: true });
    await fireEvent.keyDown(input, { key: 'a', altKey: true });
    await fireEvent.keyDown(input, { key: 'a', metaKey: true });
    await fireEvent.input(input, { value: 'new 015 test' });
    await fireEvent.paste(input, { clipboardData: { getData: vi.fn(() => 'and 89') } });
    expect(container.firstChild).toMatchSnapshot();
    vi.runAllTimers();
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('AND', expect.any(Object));
    expect(onKeyDown).toHaveBeenCalledTimes(5);
    expect(onKeyDown).toHaveBeenCalledWith('', expect.any(Object));
    expect(onPaste).toHaveBeenCalledTimes(1);
    expect(onPaste).toHaveBeenCalledWith('AND', expect.any(Object));
    rerender({
      name: 'test',
      size: 10,
      allowedKeys: { default: /z/i },
    });
    [input] = container.getElementsByTagName('input');
    await fireEvent.input(input, { value: 'zzzzzzzzzzzz' });
    await fireEvent.paste(input, { clipboardData: { getData: vi.fn(() => 'zzz') } });
    await fireEvent.input(input, { value: 'qsdqsd' });
    await fireEvent.paste(input, { clipboardData: { getData: vi.fn(() => 'sqdqsd') } });
    vi.runAllTimers();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with listeners and debounce', async () => {
    const onBlur = vi.fn();
    const onFocus = vi.fn();
    const onChange = vi.fn();
    const onIconClick = vi.fn();
    const onIconKeyDown = vi.fn();
    const { container } = render(UITextfield, {
      props: {
        name: 'test',
        icon: 'star',
        transform: (value: string): [string, number?] => [value.toUpperCase(), 1],
        size: 100,
        onBlur,
        onFocus,
        onChange,
        onIconClick,
        onIconKeyDown,
        debounceTimeout: 250,
      },
    });
    const input = container.getElementsByTagName('input')[0];
    const icon = container.getElementsByTagName('i')[0];
    await fireEvent.focus(input);
    await fireEvent.blur(input);
    await fireEvent.keyDown(icon);
    await fireEvent.click(icon);
    await fireEvent.input(input, { target: { value: 'new 015 test' } });
    vi.runAllTimers();
    await fireEvent.paste(input, { clipboardData: { getData: vi.fn(() => 'and 89 OKOK') }, target: { selectionStart: 100 } });
    vi.runAllTimers();
    expect(container.firstChild).toMatchSnapshot();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledWith('', expect.any(Object));
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledWith('', expect.any(Object));
    expect(onIconClick).toHaveBeenCalledTimes(1);
    expect(onIconClick).toHaveBeenCalledWith(expect.any(Object));
    expect(onIconKeyDown).toHaveBeenCalledTimes(1);
    expect(onIconKeyDown).toHaveBeenCalledWith(expect.any(Object));
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenCalledWith('NEW 015 TEST', expect.any(Object));
    expect(onChange).toHaveBeenCalledWith('NEW 015 TESTAND 89 OKOK', expect.any(Object));
  });
});
