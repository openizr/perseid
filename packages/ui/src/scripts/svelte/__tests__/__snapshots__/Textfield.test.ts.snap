// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svelte/UITextfield > renders correctly - autocomplete off 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - basic 1`] = `
<div>
  <div
    class="ui-textfield ui-textfield--large"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - readonly 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        readonly=""
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with autofocus 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        autofocus=""
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with disabled 1`] = `
<div>
  <div
    class="ui-textfield ui-textfield--disabled"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        disabled=""
        id="z8a8d7a5zad4"
        name="test"
        tabindex="-1"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with helper 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
    <span
      class="ui-textfield__helper"
    >
      <span
        class="ui-markdown ui-markdown--emphasis"
      >
        Helper
      </span>
    </span>
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with id 1`] = `
<div>
  <div
    class="ui-textfield"
    id="my-id"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with label 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <label
      class="ui-textfield__label"
      for="z8a8d7a5zad4"
    >
      <span
        class="ui-markdown ui-markdown--emphasis"
      >
        Label
      </span>
    </label>
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with left icon 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
      <span
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        <i
          class="ui-icon ui-icon--star"
        />
        <!--&lt;Icon&gt;-->
      </span>
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with listeners and debounce 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
      <span
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        <i
          class="ui-icon ui-icon--star"
        />
        <!--&lt;Icon&gt;-->
      </span>
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        size="100"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with placeholder 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        placeholder="test..."
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with right icon 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
      <span
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        <i
          class="ui-icon ui-icon--star"
        />
        <!--&lt;Icon&gt;-->
      </span>
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with transform and allowedKeys 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        size="10"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with transform and allowedKeys 2`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        size="10"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with type number 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        max="30"
        min="0"
        name="test"
        step="10"
        tabindex="0"
        type="number"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;

exports[`svelte/UITextfield > renders correctly - with value 1`] = `
<div>
  <div
    class="ui-textfield"
  >
     
    <div
      class="ui-textfield__wrapper"
    >
       
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        tabindex="0"
        type="text"
      />
       
    </div>
     
  </div>
  <!--&lt;Textfield&gt;-->
</div>
`;
